### PR TITLE
Fix deadlock in OvercommitTracker logging

### LINKED ISSUE #37794: Deadlock in ProcessList
https://s3.amazonaws.com/clickhouse-test-reports/37740/002f95741a1363373ee91e3dca28aff3c83c8794/stress_test__memory__actions_.html

Looks like this (however, I cannot find the root cause in the stacktraces):
```

Thread 1 (Thread 0x7f026f965d40 (LWP 665)):
#0  0x00007f026fcbe110 in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f026fcb60a3 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x000000000b5cd285 in pthread_mutex_lock ()
#3  0x0000000058e40339 in std::__1::__libcpp_mutex_lock (__m=0x723000002380) at ../contrib/libcxx/include/__threading_support:303
#4  std::__1::mutex::lock (this=0x723000002380) at ../contrib/libcxx/src/mutex.cpp:33
#5  0x000000003f028849 in std::__1::lock_guard<std::__1::mutex>::lock_guard (this=0x7ffd03ff7650, __m=...) at ../contrib/libcxx/include/__mutex_base:91
#6  DB::ProcessList::killAllQueries (this=0x723000002380) at ../src/Interpreters/ProcessList.cpp:443
#7  0x000000000b666987 in DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&)::$_10::operator()() const (this=<optimized out>) at ../programs/server/Server.cpp:1728
#8  basic_scope_guard<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&)::$_10>::invoke() (this=<optimized out>) at ../base/base/../base/scope_guard.h:99
#9  basic_scope_guard<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&)::$_10>::~basic_scope_guard() (this=this@entry=0x7ffd03ff8e50) at ../base/base/../base/scope_guard.h:48
#10 0x000000000b652b04 in DB::Server::main (this=<optimized out>) at ../programs/server/Server.cpp:1786
#11 0x00000000531a1eb6 in Poco::Util::Application::run (this=<optimized out>) at ../contrib/poco/Util/src/Application.cpp:334
#12 0x000000000b618622 in DB::Server::run (this=<optimized out>) at ../programs/server/Server.cpp:463
#13 0x00000000531fc2de in Poco::Util::ServerApplication::run (this=0x7ffd03ff9250, argc=<optimized out>, argv=0x703000014d90) at ../contrib/poco/Util/src/ServerApplication.cpp:611
#14 0x000000000b610a8f in mainEntryClickHouseServer (argc=<optimized out>, argv=0x703000014d90) at ../programs/server/Server.cpp:187
#15 0x000000000b6096fe in main (argc_=<optimized out>, argv_=0xfffffffffffffef0) at ../programs/main.cpp:439


Thread 1390 (Thread 0x7efc59bdf700 (LWP 5395)):
#0  0x00007f026fcbe110 in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f026fcb60a3 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x000000000b5cd285 in pthread_mutex_lock ()
#3  0x0000000058e40339 in std::__1::__libcpp_mutex_lock (__m=0x723000002380) at ../contrib/libcxx/include/__threading_support:303
#4  std::__1::mutex::lock (this=0x723000002380) at ../contrib/libcxx/src/mutex.cpp:33
#5  0x000000003f0141c8 in std::__1::unique_lock<std::__1::mutex>::unique_lock (this=0x7efc59bd3358, __m=...) at ../contrib/libcxx/include/__mutex_base:119
#6  DB::ProcessList::insert (this=<optimized out>, query_=..., ast=<optimized out>, query_context=...) at ../src/Interpreters/ProcessList.cpp:84
#7  0x000000003f9cc467 in DB::executeQueryImpl (begin=<optimized out>, end=<optimized out>, context=..., internal=<optimized out>, stage=<optimized out>, istr=<optimized out>) at ../src/Interpreters/executeQuery.cpp:580
#8  0x000000003f9c5ed7 in DB::executeQuery (query=..., context=..., internal=<optimized out>, stage=<optimized out>) at ../src/Interpreters/executeQuery.cpp:1069
#9  0x00000000428ac31c in DB::TCPHandler::runImpl (this=<optimized out>) at ../src/Server/TCPHandler.cpp:332
#10 0x00000000428e7e4c in DB::TCPHandler::run (this=0x71a0031d2e00) at ../src/Server/TCPHandler.cpp:1781
#11 0x000000005315aa1c in Poco::Net::TCPServerConnection::start (this=0x723000002380) at ../contrib/poco/Net/src/TCPServerConnection.cpp:43
#12 0x000000005315be66 in Poco::Net::TCPServerDispatcher::run (this=<optimized out>) at ../contrib/poco/Net/src/TCPServerDispatcher.cpp:115
#13 0x0000000053878492 in Poco::PooledThread::run (this=0x7150003d65a0) at ../contrib/poco/Foundation/src/ThreadPool.cpp:199
#14 0x00000000538738c0 in Poco::(anonymous namespace)::RunnableHolder::run (this=<optimized out>) at ../contrib/poco/Foundation/src/Thread.cpp:55
#15 0x000000005386f756 in Poco::ThreadImpl::runnableEntry (pThread=<optimized out>) at ../contrib/poco/Foundation/src/Thread_POSIX.cpp:345
#16 0x00007f026fcb3609 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#17 0x00007f026fbd8163 in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 863 (Thread 0x7efe6a0b2700 (LWP 3014)):
#0  0x00007f026fcbe110 in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f026fcb60a3 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x000000000b5cd285 in pthread_mutex_lock ()
#3  0x0000000058e40339 in std::__1::__libcpp_mutex_lock (__m=0x723000002380) at ../contrib/libcxx/include/__threading_support:303
#4  std::__1::mutex::lock (this=0x723000002380) at ../contrib/libcxx/src/mutex.cpp:33
#5  0x000000003f022ca5 in std::__1::lock_guard<std::__1::mutex>::lock_guard (this=0x7efe6a0a83b0, __m=...) at ../contrib/libcxx/include/__mutex_base:91
#6  DB::ProcessListEntry::~ProcessListEntry (this=0x703001816428) at ../src/Interpreters/ProcessList.cpp:269
#7  0x000000003f03592d in std::__1::__destroy_at<DB::ProcessListEntry, 0> (__loc=0x723000002380) at ../contrib/libcxx/include/__memory/construct_at.h:56
#8  std::__1::destroy_at<DB::ProcessListEntry, 0> (__loc=0x723000002380) at ../contrib/libcxx/include/__memory/construct_at.h:81
#9  std::__1::allocator_traits<std::__1::allocator<DB::ProcessListEntry> >::destroy<DB::ProcessListEntry, void, void> (__p=0x723000002380) at ../contrib/libcxx/include/__memory/allocator_traits.h:317
#10 std::__1::__shared_ptr_emplace<DB::ProcessListEntry, std::__1::allocator<DB::ProcessListEntry> >::__on_zero_shared (this=<optimized out>) at ../contrib/libcxx/include/__memory/shared_ptr.h:310
#11 0x000000003babde63 in std::__1::__shared_count::__release_shared (this=0x703001816410) at ../contrib/libcxx/include/__memory/shared_ptr.h:174
#12 std::__1::__shared_weak_count::__release_shared (this=0x703001816410) at ../contrib/libcxx/include/__memory/shared_ptr.h:216
#13 std::__1::shared_ptr<DB::ProcessListEntry>::~shared_ptr (this=0x7efe6a0a84b0) at ../contrib/libcxx/include/__memory/shared_ptr.h:703
#14 std::__1::shared_ptr<DB::ProcessListEntry>::reset (this=0x71a0026402c0) at ../contrib/libcxx/include/__memory/shared_ptr.h:769
#15 DB::BlockIO::reset (this=0x71a0026402c0) at ../src/QueryPipeline/BlockIO.cpp:21
#16 DB::BlockIO::operator= (this=0x71a0026402c0, rhs=...) at ../src/QueryPipeline/BlockIO.cpp:32
#17 0x00000000428f0fa1 in DB::QueryState::operator= (this=<optimized out>, this@entry=0x71a0026401e0) at ../src/Server/TCPHandler.h:46
#18 0x00000000428acac0 in DB::QueryState::reset (this=0x71a0026401e0) at ../src/Server/TCPHandler.h:110
#19 DB::TCPHandler::runImpl (this=<optimized out>) at ../src/Server/TCPHandler.cpp:393
#20 0x00000000428e7e4c in DB::TCPHandler::run (this=0x71a002640000) at ../src/Server/TCPHandler.cpp:1781
#21 0x000000005315aa1c in Poco::Net::TCPServerConnection::start (this=0x723000002380) at ../contrib/poco/Net/src/TCPServerConnection.cpp:43
#22 0x000000005315be66 in Poco::Net::TCPServerDispatcher::run (this=<optimized out>) at ../contrib/poco/Net/src/TCPServerDispatcher.cpp:115
#23 0x0000000053878492 in Poco::PooledThread::run (this=0x7150002b4d20) at ../contrib/poco/Foundation/src/ThreadPool.cpp:199
#24 0x00000000538738c0 in Poco::(anonymous namespace)::RunnableHolder::run (this=<optimized out>) at ../contrib/poco/Foundation/src/Thread.cpp:55
#25 0x000000005386f756 in Poco::ThreadImpl::runnableEntry (pThread=<optimized out>) at ../contrib/poco/Foundation/src/Thread_POSIX.cpp:345
#26 0x00007f026fcb3609 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#27 0x00007f026fbd8163 in clone () from /lib/x86_64-linux-gnu/libc.so.6

```

### PR DESCRIPTION
**Pull request by novikd**
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Do not allow recursive usage of OvercommitTracker during logging. Fixes #37794
cc @tavplubix @davenger

### CODE REVIEW

File: src/Common/OvercommitTracker.cpp [hunk 1]
```diff
@@ -23,8 +23,16 @@ OvercommitTracker::OvercommitTracker(std::mutex & global_mutex_)
     , allow_release(true)
 {}
 
+#define LOG_DEBUG_SAFE(...)                       \
+    {                                             \
+        OvercommitTrackerBlockerInThread blocker; \
+        LOG_DEBUG(__VA_ARGS__);                   \
+    }
```
Comment (davenger) [inline src/Common/OvercommitTracker.cpp:30]: You can slightly modify macro to preserve semicolons where the macro is used in the code:
```
#define LOG_DEBUG_SAFE(...)                       \
    do {                                          \
        OvercommitTrackerBlockerInThread blocker; \
        LOG_DEBUG(__VA_ARGS__);                   \
    } while (false)
```

**Comment by nikitamikhaylov** [comment]
I'd rather tried to get rid of all allocations (as much as possible) in the OvercommitMemoryTracker. Because that code will work under high memory pressure and that could be an issue when we are trying to allocate more memory when we have to free it for another query.

**Comment by novikd** [comment]
@nikitamikhaylov The problem here is that we have logging inside of the OvercommitTracker (for debug purposes). And logging may lead to the allocation of the memory. I can't get rid of these allocations (if we remove logs from here, I don't know how to debug it).
We definitely shouldn't have in call stack something like: OvercommitTracker -> Logging -> OvercommitTracker. This PR tries to do it correctly.

**Comment by tavplubix** [comment]
@novikd, there's another problem and it's not only about logging. When some code does memory allocations it may throw exceptions like `MEMORY_LIMIT_EXCEEDED`/`CANNOT_ALLOCATE_MEMORY`/`std::bad_alloc`/etc from unexpected places. It may cause obscure and hard-to-debug issues on high memory pressure if that code is not exception-safe. For example:
https://github.com/ClickHouse/ClickHouse/blob/60b9d81773d5a56486c5685ec3880d6caf207d1f/src/Common/OvercommitTracker.cpp#L65-L68

Looks like the line 68 may allocate memory and therefore throw (at least it's not clear from the code why it cannot). In this case changes to `allow_release` and `required_memory` will not be undone. Consider using `DENY_ALLOCATIONS_IN_SCOPE` or `NOEXCEPT_SCOPE` macros here to avoid such issues.

**Comment by novikd** [comment]
@tavplubix @nikitamikhaylov I removed usage of unordered_map in OvercommitTracker, and now we have no allocations there. But at some point, we'd like to have a FIFO order for thread releasing, which will highly likely require memory allocations.

**Comment by tavplubix** [comment]
> and now we have no allocations there

Let's add `DENY_ALLOCATIONS_IN_SCOPE` just to be sure

**Comment by tavplubix** [comment]
> Isn't logging itself doing allocations when formatting strings?

It is. We can use something like 

```
DENY_ALLOCATIONS_IN_SCOPE;
doCriticalStuff();
{
    ALLOW_ALLOCATIONS_IN_SCOPE;
    LOG_INFO(log, "It can do allocations");
}
continueDoingCriticalStuff();
```

Or a bit more paranoid:
```
DENY_ALLOCATIONS_IN_SCOPE;
doCriticalStuff();
{
    NOEXCEPT_SCOPE;
    /// logging is done in try-catch, so we don't expect any exceptions here
    ALLOW_ALLOCATIONS_IN_SCOPE;
    LOG_INFO(log, "It can do allocations");
}
continueDoingCriticalStuff();
```

**Comment by davenger** [comment]
Maybe do it in the macro, like this:
```
  #define LOG_DEBUG_SAFE(...)                       \
    do {                                            \
        OvercommitTrackerBlockerInThread blocker;   \
        try                                         \ 
        {                                           \
            ALLOW_ALLOCATIONS_IN_SCOPE;             \
            LOG_DEBUG(__VA_ARGS__);                 \
        }                                           \
        catch (std::badalloc &)                     \
        {                                           \
             fprintf(stderr, "Allocation failed when writing to log in OvercommitTracker\n");   \
        }                                           \
    } while (false)
```

File: src/Common/OvercommitTracker.cpp [hunk 2]
```diff
@@ -34,6 +54,8 @@ OvercommitResult OvercommitTracker::needToStopQuery(MemoryTracker * tracker, Int
     std::unique_lock<std::mutex> global_lock(global_mutex);
     std::unique_lock<std::mutex> lk(overcommit_m);
 
+    size_t id = next_id++;
```
Comment (tavplubix) [inline src/Common/OvercommitTracker.cpp:57]: Should we make `next_id` atomic and use `fetch_add`?

Comment (novikd) [reply src/Common/OvercommitTracker.cpp:57]: This code is always performed in a single thread, so there is no need to use atomic

Comment (tavplubix) [reply src/Common/OvercommitTracker.cpp:57]: This code is called from
https://github.com/ClickHouse/ClickHouse/blob/60b9d81773d5a56486c5685ec3880d6caf207d1f/src/Common/MemoryTracker.cpp#L220-L221
And it does not look like single-threaded code at all

Comment (novikd) [reply src/Common/OvercommitTracker.cpp:57]: Every public member function of OvercommitTracker starts with lock acquisition. So all code inside of OvercommitTracker is single-threaded. In `OvercommitTracker::needToStopQuery` lock acquired here:
https://github.com/ClickHouse/ClickHouse/blob/099055c1833224bf9212897a7a283a7cf529c803/src/Common/OvercommitTracker.cpp#L55

File: src/Common/OvercommitTracker.cpp [hunk 3]
```diff
@@ -20,11 +20,31 @@ OvercommitTracker::OvercommitTracker(std::mutex & global_mutex_)
...
 
+#define LOG_DEBUG_SAFE(...)                                                                    \
+    do {                                                                                       \
+        OvercommitTrackerBlockerInThread blocker;                                              \
+        try                                                                                    \
+        {                                                                                      \
+            ALLOW_ALLOCATIONS_IN_SCOPE;                                                        \
+            LOG_DEBUG(__VA_ARGS__);                                                            \
+        }                                                                                      \
+        catch (std::bad_alloc const &)                                                         \
```
Comment (tavplubix) [inline src/Common/OvercommitTracker.cpp:36]:
```suggestion
        catch (...)                                                         \
```

File: src/Common/OvercommitTracker.h [hunk 4]
```diff
@@ -7,6 +7,7 @@
 #include <cassert>
 #include <chrono>
 #include <condition_variable>
+#include <cstddef>
```
Comment (tavplubix) [inline src/Common/OvercommitTracker.h:10]: Looks redundant

**Review by tavplubix — APPROVED** [review]
LGTM

**Review by davenger — APPROVED** [review]

**Comment by novikd** [comment]
@Mergifyio update

**Comment by novikd** [comment]
@Mergifyio update

**Comment by novikd** [comment]
> PullRequestCI / TestsBugfixCheck (pull_request) Failing after 27s — TestsBugfixCheck

Ok. No tests added.

**Pull request merged**